aggregate/aggregator.go

```diff
@@ -30,52 +31,80 @@ type Aggregator struct {
 	records          []*Record
 	aggSize          int // Size of both records, and partitionKeys in bytes
 	maxAggRecordSize int
+	stringGen        *util.RandomStringGenerator
 }
 
 // NewAggregator create a new aggregator
-func NewAggregator() *Aggregator {
+func NewAggregator(stringGen *util.RandomStringGenerator) *Aggregator {
 
 	return &Aggregator{
 		partitionKeys:    make(map[string]uint64, 0),
 		records:          make([]*Record, 0),
 		maxAggRecordSize: defaultMaxAggRecordSize,
 		aggSize:          initialAggRecordSize,
+		stringGen:        stringGen,
 	}
 }
 
 // AddRecord to the aggregate buffer.
 // Will return a kinesis PutRecordsRequest once buffer is full, or if the data exceeds the aggregate limit.
-func (a *Aggregator) AddRecord(partitionKey string, data []byte) (entry *kinesis.PutRecordsRequestEntry, err error) {
+func (a *Aggregator) AddRecord(partitionKey string, hasPartitionKey bool, data []byte) (entry *kinesis.PutRecordsRequestEntry, err error) {
 
-	partitionKeySize := len([]byte(partitionKey))
-	if partitionKeySize < 1 {
-		return nil, fmt.Errorf("Invalid partition key provided")
+	if hasPartitionKey {
+		partitionKeySize := len([]byte(partitionKey))
+		if partitionKeySize < 1 {
+			return nil, fmt.Errorf("Invalid partition key provided")
+		}
 	}
 
 	dataSize := len(data)
 
 	// If this is a very large record, then don't aggregate it.
 	if dataSize >= a.maxAggRecordSize {
+		if !hasPartitionKey {
+			partitionKey = a.stringGen.RandomString()
+		}
 		return &kinesis.PutRecordsRequestEntry{
 			Data:         data,
 			PartitionKey: aws.String(partitionKey),
 		}, nil
 	}
+
+	if !hasPartitionKey {
+		if len(a.partitionKeys) > 0 {
+			// Take the first partition key from the map, if any
+			for k, _ := range a.partitionKeys {
```

Comment:
I'm open to suggestions for how to do this better, not sure how to get keys out of a map properly.

Reply:
Could I ask a question about what "first" means here? Does it mean the first element/key was put in the map? Then I am not sure if you could get it from the map through iteration.

Reply:
Could I ask a question about what "first" means here? Does it mean the first element/key was put in the map? Then I am not sure if you could get it from the map through iteration.

Reply:
I think I generally just want any key that is in the map, basically we choose a "random" key for the record but AFAICT this key does not matter, since the real partition key is just the first key, so this is just a way to save space. I'll update the comment.